### PR TITLE
ref(js): Convert FormField to a FC

### PR DESCRIPTION
**Pull request by evanpurkhiser**

### CODE REVIEW

File: static/app/components/forms/formField/index.tsx [hunk 1]
```diff
@@ -160,162 +170,140 @@ type PassthroughProps = Omit<
...
+    () => context.form ?? (new MockModel(props) as any)
+  );
 
-    this.input = node;
-  };
+  // Register field within the model
+  useEffect(() => {
+    model.setFieldDescriptor(name, initialProps.current);
+    return () => model.removeField(name);
+  }, [model, name]);
```
Comment (markstory) [inline static/app/components/forms/formField/index.tsx:190]: The old logic was only run on mount, couldn't this end-up passing stale props to `model.setFieldDescription()` should `name` change?

Comment (evanpurkhiser) [reply static/app/components/forms/formField/index.tsx:190]: Yeah it could, but that's basically the same behavior as run on mount right?

Comment (markstory) [reply static/app/components/forms/formField/index.tsx:190]: As long as the name and model never change :smile:

Comment (priscilawebdev) [reply static/app/components/forms/formField/index.tsx:190]: theoretically, the name wouldn't change after the component was mounted, right?

**Review by priscilawebdev — APPROVED** [review]

**Comment by evanpurkhiser** [comment]
Reviving this half a year old PR, I got #49525 in, so I think we can get this one in now too.

File: static/app/components/forms/formField/index.tsx [hunk 2]
```diff
@@ -341,20 +336,26 @@ class FormField extends Component<FormFieldProps> {
...
-                    const error = this.getError();
-                    const value = model.getValue(name);
+                    const error = getError();
+
+                    // XXX(epurkhiser): It is possible to use the form field in
+                    // a controlled state. THIS IS NOT WELL SUPPORTED though
+                    // and isn't really recommended.
+                    //
+                    // For now we will leave the `value` type off FormFieldProps.
+                    const value = (props as any).value ?? model.getValue(name);
```
Comment (evanpurkhiser) [inline static/app/components/forms/formField/index.tsx:346]: This was pretty nasty before.

Previously every time we rendered we re-created the MockModel object and injected the current value from the props.

**Pull request merged**